tor/controller.go

```diff
@@ -271,37 +277,57 @@ func (c *Controller) authenticate() error {
 	return nil
 }
 
-// getAuthCookie retrieves the authentication cookie in bytes from the Tor
-// server. Cookie authentication must be enabled for this to work. The boolean
-func (c *Controller) getAuthCookie() ([]byte, error) {
-	// Retrieve the authentication methods currently supported by the Tor
-	// server.
-	authMethods, cookieFilePath, version, err := c.ProtocolInfo()
+// Contains tells whether a contains x.
+func Contains(a []string, x string) bool {
+	for _, n := range a {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// authenticate authenticates the connection between the controller and the
+// Tor server using the SAFECOOKIE authentication method.
+func (c *Controller) authenticate() error {
+	protocolInfo, err := c.getProtocolInfo()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// With the version retrieved, we'll cache it now in case it needs to be
 	// used later on.
-	c.version = version
-
-	// Ensure that the Tor server supports the SAFECOOKIE authentication
-	// method or the NULL method. If NULL, we don't need the cookie info
-	// below this loop, so we just return.
-	safeCookieSupport := false
-	for _, authMethod := range authMethods {
-		if authMethod == "SAFECOOKIE" {
-			safeCookieSupport = true
-		}
-		if authMethod == "NULL" {
-			return nil, nil
-		}
+	c.version = protocolInfo.version
+
+	// NULL authentication is the simplest, use it if it's available.
+	if Contains(protocolInfo.authMethods, "NULL") {
+		return c.authenticateViaNull()
+	}
+
+	// If we have a password, and the HASHEDPASSWORD auth method is
+	// supported, then try that.
+	if Contains(protocolInfo.authMethods, "HASHEDPASSWORD") && c.password != "" {
+		return c.authenticateViaHashedPassword()
+	}
+
+	// If the SAFECOOKIE method is supported, try that.
+	if Contains(protocolInfo.authMethods, "SAFECOOKIE") {
```

Comment:
Now that we have a few different auth methods, it would be nice to have `"SAFECOOKIE"` etc declared as constants.